From 3e70969e44ee52d72053145dab2cbad74109c685 Mon Sep 17 00:00:00 2001
From: Jason Gerecke <killertofu@gmail.com>
Date: Thu, 7 Sep 2017 17:51:06 -0700
Subject: [PATCH] HID: wacom: generic: Send BTN_TOOL_PEN in prox once the pen enters range
Git-commit: 3e70969e44ee52d72053145dab2cbad74109c685
Patch-mainline: v4.15-rc1
References: bsc#1051510

When a pen is first able to to be sensed by the tablet, we would like
to inform userspace that a tool is nearby so that it can attempt to
perform palm rejection. Unfortunately, we don't know any information
about the tool that is nearby, so the best we can do is send a prox
event for a generic BTN_TOOL_PEN. If the pen later comes closer and
enters proximity, we can determine the actual tool type and send
BTN_TOOL_PEN out of prox if necessary.

Signed-off-by: Ping Cheng <ping.cheng@wacom.com>
Signed-off-by: Jason Gerecke <jason.gerecke@wacom.com>
Signed-off-by: Jiri Kosina <jkosina@suse.cz>
Acked-by: Takashi Iwai <tiwai@suse.de>

---
 drivers/hid/wacom_wac.c | 11 +++++++++++
 1 file changed, 11 insertions(+)

diff --git a/drivers/hid/wacom_wac.c b/drivers/hid/wacom_wac.c
index 9b3a247ee552..929a1ceabc21 100644
--- a/drivers/hid/wacom_wac.c
+++ b/drivers/hid/wacom_wac.c
@@ -2247,6 +2247,17 @@ static void wacom_wac_pen_report(struct hid_device *hdev,
 			wacom_wac->tool[0] = wacom_intuos_get_tool_type(wacom_wac->id[0]);
 		else
 			wacom_wac->tool[0] = BTN_TOOL_PEN;
+
+		if (wacom_wac->shared->stylus_in_proximity &&
+		    wacom_wac->tool[0] != BTN_TOOL_PEN) {
+			input_report_key(input, BTN_TOOL_PEN, 0);
+			input_sync(input);
+		}
+	}
+	else if (!wacom_wac->tool[0] && !range) { /* entering in sense */
+		input_report_key(input, BTN_TOOL_PEN, sense);
+		input_report_key(input, BTN_TOUCH, 0);
+		input_sync(input);
 	}
 
 	/* keep pen state for touch events */
-- 
2.16.4

